sources/libClangSharp/ClangSharp.cpp

```diff
@@ -54,80 +411,2151 @@ CX_CastKind clangsharp_Cursor_getCastKind(CXCursor C) {
+        }
+
+        if (const SwitchStmt* SS = dyn_cast<SwitchStmt>(S)) {
+            return MakeCXCursor(SS->getSwitchCaseList(), getCursorDecl(C), getCursorTU(C));
+        }
+
+        if (const ValueStmt* VS = dyn_cast<ValueStmt>(S)) {
+            return MakeCXCursor(VS->getExprStmt(), getCursorDecl(C), getCursorTU(C));
+        }
+    }
+
+    return clang_getNullCursor();
+}
+
+CXCursor clangsharp_Cursor_getSubExprAsWritten(CXCursor C) {
+    if (clang_isExpression(C.kind)) {
+        const Expr* E = getCursorExpr(C);
+
+        if (const CastExpr* CE = dyn_cast<CastExpr>(E)) {
+            return MakeCXCursor(CE->getSubExprAsWritten(), getCursorDecl(C), getCursorTU(C));
+        }
+    }
+
+    return clang_getNullCursor();
+}
+
+CXCursor clangsharp_Cursor_getTargetUnionField(CXCursor C) {
+    if (clang_isExpression(C.kind)) {
+        const Expr* E = getCursorExpr(C);
+
+        if (const CastExpr* CE = dyn_cast<CastExpr>(E)) {
+            return MakeCXCursor(CE->getTargetUnionField(), getCursorTU(C));
+        }
+    }
+
+    return clang_getNullCursor();
+}
+
+CXCursor clangsharp_Cursor_getTemplateArgument(CXCursor C, unsigned i) {
+    if (clang_isDeclaration(C.kind)) {
+        const Decl* D = getCursorDecl(C);
+
+        if (const ClassTemplatePartialSpecializationDecl* CTPSD = dyn_cast<ClassTemplatePartialSpecializationDecl>(D)) {
+            return MakeCXCursor(CTPSD->getTemplateParameters()->getParam(i), getCursorTU(C));
+        }
+
+        if (const TemplateDecl* TD = dyn_cast<TemplateDecl>(D)) {
+            return MakeCXCursor(TD->getTemplateParameters()->getParam(i), getCursorTU(C));
+        }
+    }
+
+    return clang_getNullCursor();
+}
+
+CXSourceLocation clangsharp_Cursor_getTemplateArgumentLocLocation(CXCursor C, unsigned i) {
+    if (clang_isDeclaration(C.kind)) {
+        const Decl* D = getCursorDecl(C);
+
+        if (const ClassScopeFunctionSpecializationDecl* CSFSD = dyn_cast<ClassScopeFunctionSpecializationDecl>(D)) {
+            SourceLocation SLoc = CSFSD->getTemplateArgsAsWritten()->arguments()[i].getLocation();
+            return translateSourceLocation(getASTUnit(getCursorTU(C))->getASTContext(), SLoc);
+        }
+
+        if (const TemplateTemplateParmDecl* TTPD = dyn_cast<TemplateTemplateParmDecl>(D)) {
+            if (i == 0) {
+                SourceLocation SLoc = TTPD->getDefaultArgument().getLocation();
+                return translateSourceLocation(getASTUnit(getCursorTU(C))->getASTContext(), SLoc);
+            }
+        }
+    }
+
+    return clang_getNullLocation();
+}
+
+CXCursor clangsharp_Cursor_getTemplateArgumentLocSourceDeclExpression(CXCursor C, unsigned i) {
+    if (clang_isDeclaration(C.kind)) {
+        const Decl* D = getCursorDecl(C);
+
+        if (const ClassScopeFunctionSpecializationDecl* CSFSD = dyn_cast<ClassScopeFunctionSpecializationDecl>(D)) {
+            const Expr* E = CSFSD->getTemplateArgsAsWritten()->arguments()[i].getSourceDeclExpression();
```

Comment:
Most of libClang proper doesn't do bounds checking, so I didn't add it here either.

Reply:
That seems reasonable then. It just stood out to me since the `i == 0` below is a bounds check of sorts (kinda but not really.)

Reply:
Those ones are specifically because I didn't want to come up with a different name for the variants that only ever have a single value, the `i` in those cases is effectively ignored 😅